### PR TITLE
[MI-2035]: Integrated unlinking project from details page

### PR DESCRIPTION
**Pull request by avas27JTG**
Integrated unlinking project from the details page

### CODE REVIEW

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/api.go [hunk 1]
```diff
@@ -148,6 +148,14 @@ func (p *Plugin) handleGetAllLinkedProjects(w http.ResponseWriter, r *http.Reque
 		return
 	}
 
+	w.Header().Add("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
```
Comment (manojmalik20) [inline server/plugin/api.go:152]:
```suggestion
```
This is the default status code if we don't write any.

File: webapp/src/components/buttons/iconButton/index.tsx [hunk 2]
```diff
@@ -22,6 +24,10 @@ const IconButton = ({tooltipText, iconClassName, extraClass = '', iconColor, onC
                 variant='outline-danger'
                 className={`plugin-btn button-wrapper btn-icon ${extraClass}`}
                 onClick={onClick}
+                aria-label={tooltipText}
+                role='button'
+                tabIndex={0}
+                onKeyDown={() => onPressingEnterKey(event, () => onClick?.())}
```
Comment (manojmalik20) [inline webapp/src/components/buttons/iconButton/index.tsx:30]:
```suggestion
                onKeyDown={(event) => onPressingEnterKey(event, () => onClick?.())}
```

File: webapp/src/components/emptyState/index.tsx [hunk 3]
```diff
@@ -15,40 +15,78 @@ type EmptyStatePropTypes = {
...
-                    <path d='M0 11.865l2.995-3.953 11.208-4.557v-3.292l9.828 7.188-20.078 3.896v10.969l-3.953-1.141zM32 5.932v19.536l-7.672 6.531-12.401-4.073v4.073l-7.974-9.885 20.078 2.396v-17.26z'/>
-                </svg>
-            </div>
-            <p className='no-data__title'>{title}</p>
-            {subTitle && (
-                <>
-                    <p className='no-data__subtitle'>{subTitle.text}</p>
+// TODO: UI to be changed
+const EmptyState = ({title, subTitle, buttonText, buttonAction, icon = 'folder'}: EmptyStatePropTypes) => {
+    return (
```
Comment (manojmalik20) [inline webapp/src/components/emptyState/index.tsx:20]:
```suggestion
const EmptyState = ({title, subTitle, buttonText, buttonAction, icon = 'folder'}: EmptyStatePropTypes) => (
```
Do this change wherever needed.

File: webapp/src/components/emptyState/index.tsx [hunk 4]
```diff
@@ -15,40 +15,78 @@ type EmptyStatePropTypes = {
...
+        <div className='no-data d-flex'>
+            <div className='d-flex flex-column align-items-center'>
+                <div className='no-data__icon d-flex justify-content-center align-items-center'>
+                    {
+                        icon === 'azure' && (
+                            <svg
```
Comment (manojmalik20) [inline webapp/src/components/emptyState/index.tsx:26]: Move this svg to a separate icons file and use here after importing.

File: webapp/src/components/emptyState/index.tsx [hunk 5]
```diff
@@ -15,40 +15,78 @@ type EmptyStatePropTypes = {
...
-                        className='plugin-btn no-data__btn btn btn-primary'
-                    >
-                        {buttonText}
-                    </button>
-                )
-            }
+                    </>
+                )}
+                {
+                    buttonText && buttonAction && (
```
Comment (manojmalik20) [inline webapp/src/components/emptyState/index.tsx:78]:
```suggestion
                {buttonText && buttonAction && (
```

File: webapp/src/components/modal/confirmationModal/index.tsx [hunk 6]
```diff
@@ -7,22 +7,25 @@ type ConfirmationModalProps = {
...
-        onHide={onHide}
-        onConfirm={onConfirm}
-        confirmAction={true}
-        confirmBtnText={confirmBtnText}
-    >
-        <p>{description}</p>
-    </Modal>
-);
+const ConfirmationModal = ({isOpen, title, confirmBtnText, description, onHide, onConfirm, isLoading}: ConfirmationModalProps) => {
+    return (
```
Comment (manojmalik20) [inline webapp/src/components/modal/confirmationModal/index.tsx:16]: Why did you add this `return` here? It was fine the way it was before.

File: webapp/src/containers/LinkModal/index.tsx [hunk 7]
```diff
@@ -50,7 +50,8 @@ const LinkModal = () => {
         setProjectDetails({...projectDetails, project: (e.target as HTMLInputElement).value});
     };
 
-    const onConfirm = useCallback(() => {
+    // Handles on confirming link project
+    const onConfirm = () => {
```
Comment (manojmalik20) [inline webapp/src/containers/LinkModal/index.tsx:54]: Why did you remove the `useCallback`?

File: webapp/src/containers/LinkModal/index.tsx [hunk 8]
```diff
@@ -71,9 +72,9 @@ const LinkModal = () => {
 
         // Make POST api request
         linkTask(projectDetails);
-    }, [errorState]);
+    };
 
-    // Make POST API request to link a project
+    // Make POST api request to link a project
```
Comment (manojmalik20) [inline webapp/src/containers/LinkModal/index.tsx:77]: Why this change?

File: webapp/src/containers/Rhs/index.tsx [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import React from 'react';
 
 import usePluginApi from 'hooks/usePluginApi';
-import {getprojectDetailsState} from 'selectors';
+import {getprojectDetailsState, getRhsState} from 'selectors';
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/index.tsx:4]: This `getRhsState` is not being used anywhere.

File: webapp/src/containers/Rhs/projectDetails/index.tsx [hunk 10]
```diff
@@ -26,31 +30,61 @@ const data: SubscriptionDetails[] = [
...
+            />
             <BackButton onClick={handleResetProjectDetails}/>
             <div className='d-flex'>
-                <p className='rhs-title'>{title}</p>
+                <p className='rhs-title'>{projectDetails.projectName}</p>
                 <IconButton
                     tooltipText='Unlink project'
                     iconClassName='fa fa-chain-broken'
                     extraClass='project-details-unlink-button unlink-button'
+                    onClick={() => handleUnlinkProject()}
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectDetails/index.tsx:87]:
```suggestion
                    onClick={handleUnlinkProject}
```

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 11]
```diff
@@ -105,6 +105,37 @@ const ProjectList = () => {
                         )
                 )
             }
+            {
+                usePlugin.getApiState(plugin_constants.pluginApiServiceConfigs.getAllLinkedProjectsList.apiServiceName).isSuccess && (
+                    data && data.length > 0 ?
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectList/index.tsx:110]:
```suggestion
                    data?.length > 0 ? (
```

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 12]
```diff
@@ -105,6 +105,37 @@ const ProjectList = () => {
...
+                        <>
+                            {
+                                data?.map((item: ProjectDetails) => (
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectList/index.tsx:113]:
```suggestion
                                data.map((item: ProjectDetails) => (
```

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 13]
```diff
@@ -105,6 +105,37 @@ const ProjectList = () => {
...
+                            }
+                            <div className='rhs-project-list-wrapper'>
+                                <button
+                                    onClick={handleOpenLinkProjectModal}
+                                    className='plugin-btn no-data__btn btn btn-primary project-list-btn'
+                                >
+                                    {'Link new project'}
+                                </button>
+                            </div>
+                        </> :
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectList/index.tsx:131]:
```suggestion
                        </>
                        ) : (
```

File: webapp/src/components/buttons/iconButton/index.tsx (hunk 2 above)
Comment (avas27JTG) [reply webapp/src/components/buttons/iconButton/index.tsx:30]: added it, but even without this it was working I guess because it was picking by default the event object

File: webapp/src/components/emptyState/index.tsx (hunk 4 above)
Comment (avas27JTG) [reply webapp/src/components/emptyState/index.tsx:26]: we have already created a `SVGWrapper` component to do this, but that is not avlb on this branch adding a TODO comment for now

File: webapp/src/containers/LinkModal/index.tsx (hunk 7 above)
Comment (avas27JTG) [reply webapp/src/containers/LinkModal/index.tsx:54]: we don't need any caching here even when the callback dependency which would be `errorState` here is changed we need to perform all the logic present inside this func

so, I don't think it worth using that, what you think?

File: webapp/src/containers/LinkModal/index.tsx (hunk 8 above)
Comment (avas27JTG) [reply webapp/src/containers/LinkModal/index.tsx:77]: this was a comment by some other reviewer on any of the PR, and I thought it's good to use caps for abbreviations

File: webapp/src/containers/Rhs/projectDetails/index.tsx (hunk 10 above)
Comment (avas27JTG) [reply webapp/src/containers/Rhs/projectDetails/index.tsx:87]: this has been fixed during refactoring in later PR

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 14]
```diff
@@ -107,7 +107,7 @@ const ProjectList = () => {
             }
             {
                 usePlugin.getApiState(plugin_constants.pluginApiServiceConfigs.getAllLinkedProjectsList.apiServiceName).isSuccess && (
-                    data && data.length > 0 ?
+                    data && data?.length > 0 ? (
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectList/index.tsx:110]:
```suggestion
                    data?.length > 0 ? (
```

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 15]
```diff
@@ -107,7 +107,7 @@ const ProjectList = () => {
...
                         <>
                             {
                                 data?.map((item: ProjectDetails) => (
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectList/index.tsx:113]:
```suggestion
                                data.map((item: ProjectDetails) => (
```

File: webapp/src/containers/LinkModal/index.tsx (hunk 7 above)
Comment (manojmalik20) [reply webapp/src/containers/LinkModal/index.tsx:54]: I don't understand. If the dependency `errorState` is not changing, then why do we need to perform all the logic inside the function?

File: webapp/src/containers/LinkModal/index.tsx (hunk 8 above)
Comment (manojmalik20) [reply webapp/src/containers/LinkModal/index.tsx:77]: I think API is better.

File: webapp/src/containers/Rhs/projectList/index.tsx (hunk 15 above)
Comment (avas27JTG) [reply webapp/src/containers/Rhs/projectList/index.tsx:113]: this is needed here as removing this gives `Object is possibly 'undefined'` it's type has `const data: void | UserDetails | ProjectDetails[] | undefined`

File: webapp/src/containers/LinkModal/index.tsx (hunk 7 above)
Comment (avas27JTG) [reply webapp/src/containers/LinkModal/index.tsx:54]: my bad, I thought the submit button stays disabled when fields are invalidated so it will be enabled only when errors are resolved and that way we would be performing the logic only when the button is clicked this implies error fields are resolved

added callback, but we can enhance our UI to keep submit button disabled if there is any error in the form and no change is made

@sooraj-shukla you can also do this as this is a good practice, it's useless to keep the button enabled when form already have validation errors

**Review by manojmalik20 — APPROVED** [review]
